### ISSUE TITLE
[Feature idea]: Image (OCI) sideload repositories

### ISSUE THREAD

**Issue by owtaylor** [flatpak/flatpak#5983]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a feature request that matches the one I want to file, without success.

### Suggestion

sideload repositories don't work for OCI remotes:

 * OCI remotes don't have collection IDs unless manually configured
 * OCI remotes have fake summary info - commit IDs in the summary are disguised manifest digests, not ostree commits - we don't know the actual ostree commits until after installation.
 * When we add signature support for OCI remotes, signatures need to be distributed in the sideload repository. Then we'll need to install from that exact signed manifest digest (or verify the result is equivalent.)

Even if we could fix all that, it would still be convenient to have sideload repositories that integrate more seemlessly to the OCI image ecosystem.

I'd like to propose a second form of sideload repository that contain OCI images. The main goal of this is using at operating system installation time along with preinstallation:

 * #5579
 * #5832

## Usage:

Create an OCI sideload repository:

```
skopeo copy docker://registry.fedoraproject.org/f41/flatpak-runtime oci:/path/to/sideload-repository
skopeo copy docker://registry.fedoraproject.org/firefox oci:/path/to/sideload-repository
```

Install from an OCI sideload repository:

```
flatpak preinstall --sideload-repo=oci:/path/to/sideload-repository
flatpak install  --sideload-repo=oci:/path/to/sideload-repository fedora firefox
```

## Implementation

See https://github.com/owtaylor/flatpak/tree/oci-sideload - not filing as a PR yet since it builds upon:

 * https://github.com/flatpak/flatpak/pull/5972

## Behavior details

Similar to other sideload repositories, a Flatpak in an image sideload repository is used for a remote, only if:

 * No image index can be fetched for a remote, or:
 * The digest of the manifest in the image sideload repository exactly matches the digest in the image index 

Image sideload repositories do not have a collection ID - they are available as a source of image for all OCI remotes

## Future: also work for ostree remotes

In the installer case, you might want to also preinstall Flatpaks from ostree remotes like Flathub. Making this work requires fixing things so that when we store an ostree commit in an OCI container we can (optionally) reconstruct the exact original commit. There is some intent that it should work but breakages include at least:

 * We don't preserve the order of the commit metadata keys
 * We add additional commit metadata keys (xa.alt-id, xa-diff-id) 
 * We don't use the parent commit ID stored in the metadata

## Future: Install from containers-storage:

Right now,  the bootc project creates an installer that has a single [containers/storage](https://github.com/containers/storage) store with the operating image and [containers to run in the operating system](https://containers.github.io/bootc/logically-bound-images.html). It would be nice to put Flatpaks into there as well. Installing could then be done by 

```
FLATPAK_SYSTEM_DIR=/mnt/sysimage/var flatpak preinstall --sideload-repo=containers-storage:
```

The difficulty here is that the format of containers-storage stores is complex and under the control of the containers/storage golang library - it's not something we can support in Flatpak. So the way we'd actually implement it is the same way that ostree does it - the skopeo tool has a `image-proxy` mode where it acts as a server and fetches images from any transport it knows about.  We could make this an optional feature of Flatpak - if you have skopeo installed you can use OCI images from any transport it knows about, if not, you just get the builtin transports (oci: / oci-archive: / docker:).

## Rationale: no collection ID

The main point of the facility is installer support - when creating an installer image, the image creator creates a OCI layout with images that the system image wants to sideload, and then at install time, they are installed from there. There’s no advantage and a lot of extra complexity if they are required to be split by collection ID.

Using a collection ID would also need a way to specify the collection ID for the collection. As a .collection-id file in the root directory? On the command line?

Collection IDs are not a security feature - if we’re offline and not checking signatures, there is no verification of sideload repositories - we’re trusting the sysadmin. When we check signatures: we check signatures against the remote we're installing into. When we are online: we check images against the summary for the remote we're installing into.

The point of collection IDs is to avoid confusion, especially in the case of permanent sideload repositories - if there is a permanently configured sideload repository of flathub content and someone does ‘flatpak install other_remote org.mozilla.Firefox’, and we’re offline and don’t have a summary for `other_remote`, we don’t want to blindly install Firefox from the Flathub sideload repository.

For now:
 * we won’t support permanently configuring an image sideload repository

In the future, collection IDs could be added for the image sideload repository, or - perhaps better - per image.

## Rationale: no index file

When I started thinking about this, I thought there would need to be an index file to map Flatpak references like `app/org.mozilla.Firefox/x86_64/stable` to the correct image, and thought about various options, all with downsides:

 * Using annotations on the OCI `index.json` entries
    * Standard tools like skopeo can’t add these annotations
    * Doesn’t work for image transports other than `oci:` / `oci-archive:`
 * [The registry index format we use for OCI remotes](https://github.com/flatpak/flatpak-oci-specs/blob/main/registry-index.md): 
    * More complex than needed
    * Maps Flatpak references to specific digests, which means it has to be regenerated specifically for the image collection.
 * A new file format (json or ini) that just maps Flatpak references to tags/references within the source. We have to define and parse a new file format - and we’d then have *two* custom image index formats used for Flatpaks.

Having no index file at all makes creation really simple as shown above - no custom tooling is needed.

**Comment by alexlarsson** [comment]
Sounds like a good idea to me. I reviewed the dependent PRs

**Comment by bam80** [comment]
Progress?

**Comment by swick** [comment]
Yes, this is supported with https://github.com/flatpak/flatpak/pull/6294.
